10.2 Домашнее задание Базам данных

tunareus opened the pull request:
1.  **Хранение в БД:**
    *   [x] **Реализовано:** Используется Room (`AppDatabase`, `LibraryItemDao`, `LibraryItemEntity`).
    *   [x] **Реализовано:** Данные сохраняются при добавлении (`repository.addItem`).
    *   [x] **Реализовано:** Данные удаляются из БД (`repository.removeItem`).
    *   [x] **Реализовано:** Данные доступны после перезапуска (обеспечивается Room).
    *   [x] **Реализовано:** Первоначальное заполнение БД при первом запуске (`LibraryDatabaseCallback`, `initializeLibraryStub`).

2.  **UI Состояния (Шиммер, Пустое состояние):**
    *   [x] **Реализовано:** Шиммер используется (`ShimmerFrameLayout` в XML, `UiState.InitialLoading` в ViewModel/Fragment).
    *   [x] **Реализовано:** Пустое состояние отображается (`emptyLayout` в XML, логика `handleEmptyState` во Fragment).
    *   [x] **Реализовано:** Минимальная длительность шиммера (1 секунда) (`MIN_SHIMMER_TIME_MS` и `delay` в ViewModel).

3.  **Сортировка:**
    *   [x] **Реализовано:** Добавлены варианты сортировки (по имени, по дате добавления) - 4 опции в Spinner (Имя А-Я/Я-А, Дата ↓/↑).
    *   [x] **Реализовано:** Логика сортировки применена к запросам в БД (`LibraryItemDao.getPagedItems` с динамическим ORDER BY).
    *   [x] **Реализовано:** Сохранение выбранной сортировки (`SettingsRepository` и `SharedPreferences`).
    *   [x] **Реализовано:** Загрузка сохраненной сортировки при запуске (в `init` ViewModel).

4.  **Пагинация:**
    *   [x] **Реализовано:** Загрузка начальной страницы (`PAGE_SIZE` в ViewModel).
    *   [x] **Реализовано:** Подгрузка следующих `N/2` элементов при скролле вниз (`loadMoreItemsBottom`, `LOAD_MORE_COUNT`).
    *   [x] **Реализовано:** Удаление `N/2` элементов *с начала* при подгрузке вниз (логика `drop` в `loadMoreItemsBottom`).
    *   [x] **Реализовано:** Подгрузка предыдущих `N/2` элементов при скролле вверх (`loadMoreItemsTop`, `LOAD_MORE_COUNT`).
    *   [x] **Реализовано:** Удаление `N/2` элементов *с конца* при подгрузке вверх (логика `dropLast` в `loadMoreItemsTop`).
    *   [x] **Реализовано:** Учет нехватки элементов при подгрузке/удалении (использование `min` в ViewModel).
    *   [x] **Реализовано:** Реализация *без* сторонних фреймворков (используется `RecyclerView.OnScrollListener` и логика в ViewModel).
    *   [x] **Реализовано:** Отображение индикаторов загрузки (ProgressBar) во время подгрузки (`AdapterItem.LoadingTop/Bottom`, `item_loading.xml`, логика в `combine` во Fragment).

5.  **Асинхронность:**
    *   [x] **Реализовано:** Все операции с БД выполняются асинхронно (`suspend` функции в DAO, `withContext(Dispatchers.IO)` в Repository, вызовы из `viewModelScope`).

tunareus committed:
dz-databases init

tunareus committed:
feat: Add utility functions, loading layout, and database entities for library items

tunareus commented:
### А так же
*   **Очистка кода:**
    *   Удалены неиспользуемые файлы, связанные с предыдущей консольной версией, хранилищем в памяти и замененными компонентами UI (например, `Shop.kt`, `Manager.kt`, `UniqueIdGenerator.kt`, `DetailActivity.kt`, `LibraryManager.kt`, `Digitizable.kt` и т.д.).
    *   Удалены неиспользуемые реализации интерфейсов (`HomeLendable`, `InLibraryUse`, `Digitizable`) из доменных моделей (`Book`, `Newspaper`, `Disk`).
    *   Очищены неиспользуемые константы в `LibraryConstants.kt`.

vladimirlisovskij commented on app/src/main/java/com/example/myapplication/data/local/mapper/LibraryItemMapper.kt:
```diff
@@ -0,0 +1,51 @@
+            id = entityId, available = this.available, name = this.name, type = ItemType.DISK, dateAdded = currentTime,
+            pages = null, author = null, issueNumber = null, monthDisplayName = null, diskType = this.getDiskType()
+        )
+        else -> throw IllegalArgumentException("Unknown LibraryItem subclass: ${this::class.java.name}")
```
можно использовать sealed class / interface чтобы убрать ветку else

vladimirlisovskij commented on app/src/main/java/com/example/myapplication/data/local/AppDatabase.kt:
```diff
@@ -0,0 +1,166 @@
+        private var INSTANCE: AppDatabase? = null
+
+        fun getDatabase(context: Context, scope: CoroutineScope): AppDatabase {
+            return INSTANCE ?: synchronized(this) {
```
похоже на дабл чек, но не хватает второй проверки

vladimirlisovskij commented on app/src/main/java/com/example/myapplication/DetailFragment.kt:
```diff
@@ -56,7 +67,10 @@ class DetailFragment : Fragment() {
 
-        viewModel = ViewModelProvider(requireActivity())[LibraryViewModel::class.java]
+        val database = AppDatabase.getDatabase(requireContext().applicationContext, lifecycleScope)
+        val repository = LibraryRepository(database.libraryItemDao())
```
было бы круто как-то отложить создание репозитория тк при повторных вызовах он не нужен и будет зря создаваться

vladimirlisovskij commented on app/src/main/java/com/example/myapplication/data/local/dao/LibraryItemDao.kt:
```diff
@@ -0,0 +1,47 @@
+
+    suspend fun getPagedItems(limit: Int, offset: Int, sortBy: String, sortOrder: String): List<LibraryItemEntity> {
+        val validSortBy = when (sortBy.lowercase()) {
+            "name" -> "name"
```
какой-то сомнительный маппинг. да и думаю лучше выполнять его не здесь

tunareus committed:
Refactor: Use raw query for paged items and improve UI

- Changed `LibraryRepository` to use a raw SQL query for fetching paged items, enhancing performance and flexibility.
- Moved the logic for building the paged item query to `LibraryRepository`.
- Removed unused `getPagedItems` method from `LibraryItemDao`
- Added new dimens for library item elevations for better visual feedback.
- Refactored `LibraryItem` to `sealed` class
- Updated `LibraryItemAdapter` to use elevation resources for card views.
- Fixed creating database instance.
- Removed unused mapper exceptions.
- Fixed passing data to `DetailFragment`.

tunareus committed:
Reduce LOAD_MORE_COUNT to 8 and PREFETCH_DISTANCE to 5 in LibraryViewModel.
